### PR TITLE
Fixes #118 - [Performance] Excel-like filter for huge data sets

### LINKED ISSUE #118: [Performance] Excel-like filter for huge data sets
There are several places in the code that is used for the Excel-like filter dropdowns, that contain possible performance flaws for huge datasets.

- `NatCombo`  
  - `setDropdownSelection(String[])` - inside a for-loop a `contains()` operation is executed to calculated the `selectionStateMap` and the checkbox state. This should be replaced with a `Map` based operation.
  - `getSelectionIndices()` - inside a for-loop an `indexOf()` operation is executed to calculated the selected indexes based on the items. This should be hcanged to a `Map` based operation.
  - The `Integer` usage should be replaced with primitive collection usage.
- `FilterNatCombo.getTransformedText(String[])`  
  Use a `StringJoiner` instead of `+` operator for String concatenation
- `ComboBoxGlazedListsFilterStrategy.getStringFromColumnObject(int, Object)`  
  Use a `StringJoiner` instead of `+` operator for String concatenation
- `FilterRowComboBoxCellEditor.setCanonicalValue(Object)`  
  The calculation of `diffValues` is using `removeAll()` to calculate the diff between currently visible and all available items. To improve the performance we should change the type of `diffValues` from `List` to `HashSet` as the order doesn't seem to be relevant.

### PR DESCRIPTION
**Pull request by fipro78**

### CODE REVIEW

File: org.eclipse.nebula.widgets.nattable.core/src/org/eclipse/nebula/widgets/nattable/widget/NatCombo.java [hunk 1]
```diff
@@ -1273,7 +1293,7 @@ public void select(int index) {
     public void select(int[] indices) {
         if (!getDropdownTable().isDisposed()) {
             getDropdownTable().select(indices);
-            List<Integer> indicesList = ArrayUtil.asIntegerList(indices);
```
[the remark this answers is not in the document]
Comment (fipro78) [reply org.eclipse.nebula.widgets.nattable.core/src/org/eclipse/nebula/widgets/nattable/widget/NatCombo.java:1276]: Switched from IntList to IntSet. Thanks for the pointer.

[the remark this answers is not in the document]
Comment (mmnze) [reply org.eclipse.nebula.widgets.nattable.core/src/org/eclipse/nebula/widgets/nattable/widget/NatCombo.java:1276]: my English...  
Thanks for the update

File: org.eclipse.nebula.widgets.nattable.core/src/org/eclipse/nebula/widgets/nattable/filterrow/combobox/FilterRowComboBoxCellEditor.java [hunk 2]
```diff
@@ -264,7 +264,7 @@ && getComboBoxDataProvider() instanceof FilterRowComboBoxDataProvider
             // available items
             List<?> allValues = ((FilterRowComboBoxDataProvider) getComboBoxDataProvider()).getAllValues(getColumnIndex());
             List<?> visibleValues = getComboBoxDataProvider().getValues(getColumnIndex(), getRowIndex());
-            List<?> diffValues = new ArrayList<>(allValues);
+            HashSet<?> diffValues = new HashSet<>(allValues);
             diffValues.removeAll(visibleValues);
```
Comment (mmnze) [inline org.eclipse.nebula.widgets.nattable.core/src/org/eclipse/nebula/widgets/nattable/filterrow/combobox/FilterRowComboBoxCellEditor.java:268]: I just realized: This does not actually solve the issue:   
The expensive `contains` operation within the `removeAll` is not executed on "this" (ie. `diffValues`) but the method argument, an arraylist in this case. See `HashSet`, l. 175 (Java 11).  
Sorry, my bad.

**Review by mmnze — COMMENTED** [review]
And I found a new one, I'd like to add: 
`ComboBoxFilterUtils.isAllSelected`, l. 69. Temporarily putting the `dataCollection` into a set should do the trick (`containsAll` calls contains on `this`, not the provided collection).

**Comment by mmnze** [comment]
And some more (I just searched "systematically" through the code looking for potentially expensive collection operations in the filter code instead of stopping a running program if it took too long ;-) ): 

- org.eclipse.nebula.widgets.nattable.filterrow.combobox.FilterRowComboBoxCellEditor.setCanonicalValue(Object), l. 273ff (after the code we already talked about). The size of the collections involved should be much smaller here, but we probably can apply the "List to Set"-performance fix here as well
- org.eclipse.nebula.widgets.nattable.filterrow.combobox.FilterRowComboBoxDataProvider.doCommand(ILayer, UpdateDataCommand), l. 1085, 1094, 1095; probably the same idea (
- org.eclipse.nebula.widgets.nattable.extension.glazedlists.filterrow.ComboBoxFilterRowHeaderComposite.handleEvent(FilterRowComboUpdateEvent), l. 907ff
- org.eclipse.nebula.widgets.nattable.filterrow.combobox.FilterRowComboBoxDataProvider.buildUpdateEvent(FilterRowComboUpdateEvent, int, List<?>, List<?>), l. 764ff

File: org.eclipse.nebula.widgets.nattable.core/src/org/eclipse/nebula/widgets/nattable/filterrow/combobox/FilterRowComboBoxCellEditor.java (hunk 2 above)
Comment (fipro78) [reply org.eclipse.nebula.widgets.nattable.core/src/org/eclipse/nebula/widgets/nattable/filterrow/combobox/FilterRowComboBoxCellEditor.java:268]: Before I changed from `ArrayList` to `HashSet` I tested locally with this code:

```java
    @Test
    void testListPerformance() {
        ArrayList<String> all = new ArrayList<>();
        ArrayList<String> flanders = new ArrayList<>();

        for (int i = 0; i < 100_000; i++) {
            all.add("Simpson_" + i);
            all.add("Flanders_" + i);
            flanders.add("Flanders_" + i);
        }

        long start = System.currentTimeMillis();

        ArrayList<String> diff = new ArrayList<>(all);
        diff.removeAll(flanders);

        long end = System.currentTimeMillis();

        System.out.println("ArrayList#removeAll(ArrayList) - " + (end - start));

        start = System.currentTimeMillis();

        HashSet<String> diffSet = new HashSet<>(all);
        diffSet.removeAll(flanders);

        end = System.currentTimeMillis();

        System.out.println("HashSet#removeAll(ArrayList) - " + (end - start));

        start = System.currentTimeMillis();

        HashSet<String> diffSetSet = new HashSet<>(all);
        diffSetSet.removeAll(new HashSet<>(flanders));

        end = System.currentTimeMillis();

        System.out.println("HashSet#removeAll(HashSet) - " + (end - start));
    }
```

If `diff` is a List, the operation takes about 60s. The `diffSet` variant takes around 30ms. I would say that the test is similar to the code in discussion. Or what am I missing?

Comment (fipro78) [reply org.eclipse.nebula.widgets.nattable.core/src/org/eclipse/nebula/widgets/nattable/filterrow/combobox/FilterRowComboBoxCellEditor.java:268]: OK, I found out why my test was working. It strongly depends on the size of the collection passed as parameter. As long as the collection is smaller than the "this" collection, things work fast. If the collection is bigger, the performance is a mess.

I adjusted the collection creation to verify it:

```java
        for (int i = 0; i < 100_000; i++) {
            all.add("Simpson_" + i);
            all.add("Flanders_" + i);
            flanders.add("Flanders_" + i);
            flanders.add("FlandersO_" + i);
        }

        for (int i = 0; i < 10_000; i++) {
            flanders.add("FlandersXxx_" + i);
        }
```

Now the _HashSet#removeAll(ArrayList)_ takes 123s while _HashSet#removeAll(HashSet)_ is still at about 30ms.

**Pull request merged**